virtool/users/data.py

```diff
@@ -292,91 +362,265 @@ async def update(self, user_id: str, data: UpdateUserRequest):
+                timestamp = virtool.utils.timestamp()
+                updates.update(
+                    {
+                        "password": hashedpass,
+                        "last_password_change": timestamp,
+                        "invalidate_sessions": True,
+                    }
+                )
+                user.password = hashedpass
+                user.last_password_change = timestamp
+                user.invalidate_sessions = True
+
+            if "groups" in data:
+                try:
+                    updates.update(
+                        await compose_groups_update(self._pg, data["groups"])
+                    )
+
+                    await pg_session.execute(
+                        delete(user_group_associations)
+                        .where(user_group_associations.c.user_id == user.id)
+                        .where(
+                            user_group_associations.c.group_id.in_(
+                                (
+                                    await pg_session.execute(
+                                        select(
+                                            user_group_associations.c.group_id
+                                        ).where(
+                                            user_group_associations.c.user_id == user.id
+                                        )
+                                    )
+                                )
+                                .scalars()
+                                .all()
+                            )
+                        )
+                    )
+
+                    if len(data["groups"]) > 0:
+                        await pg_session.execute(
+                            insert(user_group_associations).values(
+                                [(user.id, _id) for _id in data["groups"]]
+                            )
+                        )
+
+                except DatabaseError as err:
+                    raise ResourceConflictError(str(err))
+
+            if "primary_group" in data:
+                try:
+                    primary_group = await compose_primary_group_update(
+                        self._mongo,
+                        self._pg,
+                        data.get("groups", []),
+                        data["primary_group"],
+                        user_id,
+                    )
+
+                    await pg_session.execute(
+                        delete(user_group_associations)
+                        .where(user_group_associations.c.user_id == user.id)
+                        .where(
+                            user_group_associations.c.group_id == user.primary_group_id
+                        )
+                    )
+
+                    await pg_session.execute(
+                        insert(user_group_associations).values(
+                            user_id=user.id, group_id=data["primary_group"]
+                        )
+                    )
```

Comment:
I believe correct behaviour is to error if the user is not currently or won't be a member of the group selected as the primary group

Reply:
I'm confused , this is the update function so primary group may be being changed?

Reply:
nvm understood now